sw/device/lib/dif/dif_spi_device.h

```diff
@@ -0,0 +1,233 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_SPI_DEVICE_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_SPI_DEVICE_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+/**
+ * Represents a signal edge type.
+ */
+typedef enum dif_spi_device_edge {
+  kDifSpiDeviceEdgePositive,
+  kDifSpiDeviceEdgeNegative,
+} dif_spi_device_edge_t;
+
+/**
+ * Represents a signal edge type.
+ */
+typedef enum dif_spi_device_bit_order {
+  kDifSpiDeviceBitOrderMsbToLsb,
+  kDifSpiDeviceBitOrderLsbToMsb,
+} dif_spi_device_bit_order_t;
+
+typedef enum dif_spi_device_irq_state {
+  kDifSpiDeviceIrqStateEnable = true,
+  kDifSpiDeviceIrqStateDisable = false,
+} dif_spi_device_irq_state_t;
+
+/**
+ * Represents the types of interrupts that the SPI device will
+ * fire to signal state.
+ */
+typedef enum dif_spi_device_irq_type {
+  kDifSpiDeviceIrqTypeRxFull,
+  kDifSpiDeviceIrqTypeRxAboveLevel,
+  kDifSpiDeviceIrqTypeTxBelowLevel,
+  kDifSpiDeviceIrqTypeRxError,
+  kDifSpiDeviceIrqTypeRxOverflow,
+  kDifSpiDeviceIrqTypeTxUnderflow,
+} dif_spi_device_irq_type_t;
+
+/**
+ * Errors returned by SPI DIF functions.
+ */
+typedef enum dif_spi_device_result {
```

Comment:
I think it could be useful to have more error codes for some functions.

For example, `kDifSpiDeviceResultTxFifoLenInvalid` or similar would be more descriptive.
```
  uint16_t rx_fifo_start = 0x0;
  uint16_t rx_fifo_end = config.rx_fifo_len - 1;
  uint16_t tx_fifo_start = config.rx_fifo_len;
  uint16_t tx_fifo_end = config.rx_fifo_len + config.tx_fifo_len - 1;
  if (tx_fifo_end >= kDifSpiDeviceSramLen) {
    // We've overflown the SRAM region...                                                                                                                                                      
    return kDifSpiDeviceResultBadArg;
  }
```

Reply:
What would a caller do in response to those error codes? I think that the distinction is just noise.

Reply:
In that case there wasn't much point moving away from single `bool` return to enum error codes.

Reply:
It allows the caller to at least produce a meaningful debug message, in general there probably are more use-cases that is hard to foresee beforehand.

Reply:
Sure, that makes sense.